# Assignment a-4 - Digits

### Jeff Scanlon
#### jscanlo2

### Daniel Schnelbach
#### dschnelb

In [28]:
%run -i read_all_data.py

In [29]:
train_img, train_lbl = read_all_image_label('train_images.txt','train_labels.txt')

In [30]:
test_img, test_lbl = read_all_image_label('test_images.txt','test_labels.txt')

In [31]:
print( (len(train_img), len(test_img)) )
print( (len(train_lbl),len(test_lbl)) )

(5000, 1000)
(5000, 1000)


In [32]:
# Get arrays
import numpy as np
y_train = np.array(train_lbl)
y_test = np.array(test_lbl)

In [33]:
#img = x_train[0]
img = train_img[0]

In [10]:
from pprint import pprint as pp
pp(img)

['                            ',
 '                            ',
 '                            ',
 '                            ',
 '                            ',
 '                +++++##+    ',
 '        +++++######+###+    ',
 '       +##########+++++     ',
 '        #######+##          ',
 '        +++###  ++          ',
 '           +#+              ',
 '           +#+              ',
 '            +#+             ',
 '            +##++           ',
 '             +###++         ',
 '              ++##++        ',
 '                +##+        ',
 '                 ###+       ',
 '              +++###        ',
 '            ++#####+        ',
 '          ++######+         ',
 '        ++######+           ',
 '       +######+             ',
 '    ++######+               ',
 '    +####++                 ',
 '                            ',
 '                            ',
 '                            ']


In [34]:
# encode a "digit" 

def char2num(c):
    if c == ' ':
        n = 0
    elif c == '+':
        n = 1
    elif c == '#':
        n = 2
    else:
        print('This should not happen!')
    return n

In [12]:
def img2array(img):
    s = ''.join(img)
    arr = [ char2num(c) for c in s]
    return np.array(arr)

In [13]:
arr = img2array(img)

In [160]:
arr.shape

(784,)

In [14]:
x_train = [img2array(img) for img in train_img]
x_test = [img2array(img) for img in test_img]

In [15]:
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import BernoulliNB

In [16]:
g_model = GaussianNB()
m_model = MultinomialNB()
b_model = BernoulliNB()

## Gaussian

In [252]:
from sklearn.model_selection import train_test_split

g_model.fit(x_train, y_train)
predicted = g_model.predict(x_test)
expected = y_test

In [253]:
# Import  metrics
from sklearn import metrics

# summarize the fit of the model

print(metrics.accuracy_score(expected, predicted))
print(metrics.classification_report(expected, predicted))
print(metrics.confusion_matrix(expected, predicted))

0.509
              precision    recall  f1-score   support

           0       0.78      0.87      0.82        90
           1       0.58      0.98      0.73       108
           2       0.77      0.26      0.39       103
           3       0.64      0.27      0.38       100
           4       0.62      0.17      0.26       107
           5       0.50      0.08      0.13        92
           6       0.58      0.76      0.66        91
           7       0.77      0.25      0.38       106
           8       0.29      0.59      0.39       103
           9       0.38      0.89      0.53       100

    accuracy                           0.51      1000
   macro avg       0.59      0.51      0.47      1000
weighted avg       0.59      0.51      0.47      1000

[[ 78   0   0   0   1   1   4   0   6   0]
 [  0 106   0   0   0   0   1   0   1   0]
 [  4  11  27  10   3   0  27   1  19   1]
 [  3  13   1  27   0   1   6   2  37  10]
 [  3   3   2   1  18   1   6   0  22  51]
 [  9   8   2   1   

## Gaussian performance

The Gaussian model performed quite poorly. Total accuracy is just 0.51 and the model achieved a recall > 0.8 on just three digits. Digit 5 performed worst, followed by digit 4. Each of these are commonly misclassified as 8 or 9, which is intuitive given their shapes and humans tendency to not quite close our handwritten 8s, or to close too much the tops of our 4s, etc., when we are writing quickly (my handwritten digits certainly aren't always the cleanest!). 

The wiki article suggests Gaussian models work best when we have continuous variables and can safely assume a normal distribution. This is not a good set of assumptions for our data. 

In [63]:
print(predicted[:30])
print(expected[:30])

[9 0 1 8 1 9 9 1 1 0 9 1 9 9 8 1 0 2 8 1 8 7 9 0 8 1 9 1 5 1]
[9 0 2 5 1 9 7 8 1 0 4 1 7 9 6 4 2 6 8 1 3 7 5 4 4 1 8 1 3 8]


In [73]:
pp(test_lbl[3])
pp(test_img[3])

pp(test_lbl[22])
pp(test_img[22])

5
['                            ',
 '                            ',
 '                            ',
 '                            ',
 '                            ',
 '          +#########+       ',
 '         +###########+      ',
 '         ############+      ',
 '         ############       ',
 '         ####+++#####       ',
 '         +##+     +++       ',
 '         +###++++           ',
 '          ########+         ',
 '          #########+        ',
 '          ##########+       ',
 '          +##########       ',
 '           +++  ++###+      ',
 '                  +###      ',
 '      ++           ###      ',
 '     +###++       +###      ',
 '      ######++   +###+      ',
 '      ++#############       ',
 '       ++###########+       ',
 '         +#########+        ',
 '           ++####++         ',
 '                            ',
 '                            ',
 '                            ']
5
['                            ',
 '                            ',
 '    

Above we see cases where we got 8 but expected 5. The tops of the fives each appear rather closed, and it is possible many written 8s in the data include unclosed portions aking to a 5 (i.e., not a perfectly enclosed symbol). 

In [77]:
pp(test_lbl[10])
pp(test_img[10])

pp(test_lbl[13])
pp(test_img[13])

4
['                            ',
 '                            ',
 '                            ',
 '                            ',
 '                            ',
 '                            ',
 '        +##+        ++      ',
 '       +###+        +#+     ',
 '     ++####+       +###     ',
 '     +####        +###+     ',
 '     ####+        +###+     ',
 '    +####+        +###+     ',
 '    +#####+++    +###+      ',
 '     +#######++++###+       ',
 '     +++############+       ',
 '        +++++++#####+       ',
 '              +#####+       ',
 '              +####+        ',
 '              +###+         ',
 '              +###+         ',
 '             +###+          ',
 '            +####           ',
 '            +###+           ',
 '            +###+           ',
 '            +###+           ',
 '            +##+            ',
 '                            ',
 '                            ']
9
['                            ',
 '                            ',
 '    

Above are 1) a case where we expected 4 and got 9, followed by 2) a case where we predicted and got 9. 

## Multinomial

In [162]:
m_model.fit(x_train, y_train)
predicted = m_model.predict(x_test)
expected = y_test

In [163]:
print(metrics.accuracy_score(expected, predicted))
print(metrics.classification_report(expected, predicted))
print(metrics.confusion_matrix(expected, predicted))

0.751
              precision    recall  f1-score   support

           0       0.92      0.87      0.89        90
           1       0.84      0.94      0.89       108
           2       0.86      0.73      0.79       103
           3       0.70      0.83      0.76       100
           4       0.74      0.65      0.70       107
           5       0.80      0.55      0.65        92
           6       0.79      0.78      0.78        91
           7       0.88      0.70      0.78       106
           8       0.60      0.63      0.62       103
           9       0.55      0.82      0.66       100

    accuracy                           0.75      1000
   macro avg       0.77      0.75      0.75      1000
weighted avg       0.77      0.75      0.75      1000

[[ 78   0   1   1   0   2   3   0   5   0]
 [  0 102   0   0   0   0   1   0   5   0]
 [  1   3  75   3   2   1   7   1  10   0]
 [  0   1   0  83   0   2   2   5   1   6]
 [  0   0   2   0  70   1   4   0   2  28]
 [  2   1   1  15   

## Multinomial performance

The Multinomial model is a significant improvement. Total accuracy is just 0.75 and the model achieved a recall > 0.8 on four digits. Digits 5 and 8 caused the most trouble, which these often being mistaken for 3s now. 

Multinomial models employ feature vectors, which is what we have here. This has greatly improved predictions relative to the Gaussian because it will create a histogram based on actual counting of events as they are observed instead of using a normality assumption. The model is a better fit for our feature as constructed. 

In [83]:
print(predicted[:30])
print(expected[:30])

[9 0 2 3 1 9 7 8 1 0 4 1 9 9 4 9 2 2 8 1 3 7 9 4 8 1 8 1 5 8]
[9 0 2 5 1 9 7 8 1 0 4 1 7 9 6 4 2 6 8 1 3 7 5 4 4 1 8 1 3 8]


In [93]:
pp(test_lbl[3])
pp(predicted[3])
pp(test_img[3])

pp(test_lbl[28])
pp(predicted[28])
pp(test_img[28])

5
3
['                            ',
 '                            ',
 '                            ',
 '                            ',
 '                            ',
 '          +#########+       ',
 '         +###########+      ',
 '         ############+      ',
 '         ############       ',
 '         ####+++#####       ',
 '         +##+     +++       ',
 '         +###++++           ',
 '          ########+         ',
 '          #########+        ',
 '          ##########+       ',
 '          +##########       ',
 '           +++  ++###+      ',
 '                  +###      ',
 '      ++           ###      ',
 '     +###++       +###      ',
 '      ######++   +###+      ',
 '      ++#############       ',
 '       ++###########+       ',
 '         +#########+        ',
 '           ++####++         ',
 '                            ',
 '                            ',
 '                            ']
3
5
['                            ',
 '                            ',
 '

The multinomial model confused more 5s and 3s. Above are two instances. First, calling a 5 a 3, and then a 3 a 5. 

## Bernoulli

In [164]:
b_model.fit(x_train, y_train)
predicted = b_model.predict(x_test)
expected = y_test

In [96]:
print(metrics.accuracy_score(expected, predicted))
print(metrics.classification_report(expected, predicted))
print(metrics.confusion_matrix(expected, predicted))

0.771
              precision    recall  f1-score   support

           0       0.92      0.84      0.88        90
           1       0.83      0.96      0.89       108
           2       0.85      0.78      0.81       103
           3       0.71      0.79      0.75       100
           4       0.78      0.77      0.77       107
           5       0.73      0.67      0.70        92
           6       0.81      0.76      0.78        91
           7       0.87      0.73      0.79       106
           8       0.75      0.60      0.67       103
           9       0.58      0.80      0.67       100

    accuracy                           0.77      1000
   macro avg       0.78      0.77      0.77      1000
weighted avg       0.78      0.77      0.77      1000

[[ 76   0   1   0   1   5   3   0   4   0]
 [  0 104   1   0   0   2   1   0   0   0]
 [  1   3  80   4   1   0   6   1   5   2]
 [  0   2   0  79   0   3   2   6   2   6]
 [  0   1   0   0  82   0   3   1   2  18]
 [  2   2   1  12   

## Bernoulli performance

The Bernoulli model shows marginal improvement (accuracy rises by 0.02) over the Multinomial. The difference in these models is that in the Bernoulli, binary term occurrence features are used rather than term frequencies (that make the histograms noted above). 

The performance improves mostly from better classification of 4, 5 and 8. 

## Adding features

In [52]:
def img2array2(img, val1, val2, val3):
    s = ''.join(img)
    arr = [ char2num(c) for c in s ]
    
    # Add the means for each column vector in the image
    col_means = [ x.mean() for x in np.array(arr).reshape(28,28) ]

        # Using a matrix to get row means and adding these made it WORSE!
    #arr3 = np.matrix(np.array(arr).reshape(28,28)).mean(0)
    #row_means = [ arr3[0,i] for i in range(0,arr3.shape[1]) ]


    # Put the 28x28 image into 3x3 blocks and get their aggregates
    x = list(range(0,27,3))
    y = list(range(3,28,3))

    mtx = np.matrix(np.array(arr).reshape(28,28))
    
    # ZIP(xy) pairs the 0:3 range splits above, then I use them iteratively to get the 3x3 blocks out
    block_sums = [mtx[i[0]:i[1], mtx[i[0]:i[1]]].sum() for i in zip(x,y)]
    
                
    for mu in col_means:
        arr.append(mu)
        
    #for mu in row_means:
    #    arr.append(mu)
    
    for mu in block_sums:
        arr.append(mu)
        
    for i in val1:
        arr.append(i)
    for i in val2:
        arr.append(i)
    for i in val3:
        arr.append(i)
    
    return np.array(arr)

In [36]:
arr = [ char2num(c) for c in s ]
arr;

NameError: name 's' is not defined

## Retrain

In [37]:
x_train_2 = [img2array2(img) for img in train_img]
x_test_2 = [img2array2(img) for img in test_img]

TypeError: img2array2() missing 3 required positional arguments: 'val1', 'val2', and 'val3'

In [345]:
x_train_2[0][785]

0.0

## Test with just adding my lone successful addition of the basic numpy vector means

In [416]:
g_model.fit(x_train_2, y_train)
predicted = g_model.predict(x_test_2)
expected = y_test

In [417]:
print(metrics.accuracy_score(expected, predicted))
print(metrics.classification_report(expected, predicted))
print(metrics.confusion_matrix(expected, predicted))

0.534
              precision    recall  f1-score   support

           0       0.83      0.87      0.85        90
           1       0.64      0.97      0.77       108
           2       0.77      0.32      0.45       103
           3       0.69      0.27      0.39       100
           4       0.70      0.21      0.33       107
           5       0.53      0.10      0.17        92
           6       0.61      0.78      0.69        91
           7       0.78      0.27      0.41       106
           8       0.30      0.67      0.42       103
           9       0.39      0.90      0.55       100

    accuracy                           0.53      1000
   macro avg       0.62      0.54      0.50      1000
weighted avg       0.62      0.53      0.50      1000

[[ 78   0   1   0   0   1   3   0   7   0]
 [  0 105   0   0   0   0   2   0   1   0]
 [  4  10  33   7   3   0  25   1  20   0]
 [  2  10   1  27   0   1   6   2  41  10]
 [  0   2   2   1  23   1   6   0  22  50]
 [  8   5   2   1   

### Number of bands of characters, average band length

### Note to self: nbands is not a good one to use! Drastically decreased performance somehow??
### Also, while av_length and Daniel's mean (arr2) both slightly increase performance, together they decrease performance! Weird!

In [41]:

nbands = []
av_length = []
overall_av_len = []

for X in train_img:
    num_bands = []
    avg_band_length = []
    for line in X:
        string = line.split()
        num = len(string)
        num_bands.append(num)
        if num == 0:
            avg_band_length.append(0)
        elif num == 1:
            avg_band_length.append(len(string[0]))
        else:
            for y in string:
                ans = [len(y) for y in string]
            avg_band_length.append(sum(ans)/len(ans))
    nbands.append(num_bands)
    av_length.append(avg_band_length)

### Surprisingly to me, none of these seemed to have great impact individually, and some actually decreased performance slightly.

### Nbands worked the best out of these four, but only increased .001 (.1%)!

In [42]:

oneband = []
twobands = []
threebands = []
nbands = []
for X in train_img:
    band1 = []
    band2 = []
    band3 = []
    bandmore = []
    for line in X:
        string = line.split()
        num = len(string)
        if num == 1:
            band1.append(1)
        elif num == 2:
            band2.append(1)
        elif num == 3:
            band3.append(1)
        elif num >= 4:
            bandmore.append(1)
    oneband.append([sum(band1)])
    twobands.append([sum(band2)])
    threebands.append([sum(band3)])
    nbands.append([sum(bandmore)])

### Want to try an interation of the work above...but check to see if 2 band regions are enclosed by 1 band regions on both sides or just on one side

### It kind of surprises me that this one didn't do as well because when I checked for myself, this successfully identified 7 of the 8 eights that are in the first 100 digits, and made 2 false negatives. (Or by tweaking this code you could reduce false positives to 1 but decrease true positives to 6 out of 8, by removing the "+" signs from each string).

In [48]:
import re

tot_list = []
eights = []
fours = []
for X in train_img:
    len_list = []
    for line in X:
        string = line.split()
        num = len(string)
        len_list.append(num)
    tot_list.append(len_list)
for x in tot_list:
    string = str(x).strip('[]')
    string = string.replace(',','')  
    string = string.replace(' ','')
    match = re.search("[1]+2{2,8}[1]+2{3,8}[1]+", string)
    if boolean == bool(match):
        eights.append([0])
    else:
        eights.append([8])
        
#Used this code to check my work: (Uncomment to use)
#list(enumerate(eights[0:100]))             #<- To know which ones I classify as 8 in first 100
#list(enumerate(y_train_np[0:100]))        <- To know where all the eights in first 100 are
#X_train[85]          <- Just for looking at the 8's I missed first time around
#tot_list[94]

NameError: name 'boolean' is not defined

### Let's try something similar with fours.
### This one had a lower success rate.

In [ ]:
four_list = []
fours = []
for X in train_img:
    four_list = []
    for line in X:
        string = line.split()
        num = len(string)
        len_list.append(num)
    fours.append(four_list)
for x in fours:
    string = str(x).strip('[]')
    string = string.replace(',','')  
    string = string.replace(' ','')
    match1 = re.search("[0]+2{4,8}[1]+", string)
    if boolean == bool(match1):
        fours.append([0])
    else:
        fours.append([8])
list(enumerate(fours[0:100]))
#list(enumerate(y_train_np[0:100]))

### num12 = Number of character bands of 12 or more characters
### num3 = Number of character bands of exactly 3 characters
### Seems to be somewhat effective, even with the Means

In [49]:
num12 = []
num3 = []
for X in train_img:
    counts12 = []
    counts3 = []
    for line in X:
        count12 = 0
        count3 = 0
        string = line.split()
        for s in string:
            if len(s) >= 12:
                count12 += 1
            elif len(s) == 3:
                count3 +=1
        counts12.append(count12)
        counts3.append(count3)
    num12.append([sum(counts12)])
    num3.append([sum(counts3)])

In [50]:
def img2array(img, val1, val2, val3):
    s = ''.join(img)
    arr = [ char2num(c) for c in s]
    arr2 = [ x.mean() for x in np.array(arr).reshape(28,28) ]
    for i in arr2:
        arr.append(i)
    for i in val1:
        arr.append(i)
    for i in val2:
        arr.append(i)
    for i in val3:
        arr.append(i)
    return np.array(arr)

In [53]:
X_train_np = [img2array(img, val1, val2, val3) for img, val1, val2, val3 in zip(train_img, num12, num3, nbands)]
X_test_np = [img2array(img, val1, val2, val3) for img, val1, val2, val3 in zip(test_img, num12, num3, nbands)]

In [57]:
g_model.fit(X_train_np, y_train)
predicted = g_model.predict(X_test_np)
expected = y_test

In [59]:
from sklearn import metrics

print(metrics.accuracy_score(expected, predicted))
print(metrics.classification_report(expected, predicted))
print(metrics.confusion_matrix(expected, predicted))

0.546
              precision    recall  f1-score   support

           0       0.83      0.87      0.85        90
           1       0.63      0.98      0.77       108
           2       0.77      0.33      0.46       103
           3       0.71      0.34      0.46       100
           4       0.72      0.21      0.33       107
           5       0.56      0.10      0.17        92
           6       0.62      0.78      0.69        91
           7       0.78      0.29      0.42       106
           8       0.33      0.68      0.44       103
           9       0.39      0.90      0.55       100

    accuracy                           0.55      1000
   macro avg       0.63      0.55      0.51      1000
weighted avg       0.63      0.55      0.51      1000

[[ 78   0   1   0   0   1   3   0   7   0]
 [  0 106   0   0   0   0   1   0   1   0]
 [  4  10  34   8   3   0  24   1  18   1]
 [  2  11   1  34   0   1   6   2  33  10]
 [  0   3   2   1  23   1   7   1  20  49]
 [  8   7   2   1   

### We were able to boost performance to 0.546...a very slight boost and not quite what we were hoping for, but it is still improvement! Below are some additional ideas for what we would like to pursue next if we had more time.

### Other features we considered but were not sure how to implement:
Feautres involving columns.
Examples: Max length of consecutive 1's in any column in the 28x28 array.
Max length of consecutive 2's in any column...
Max length of consecutive 1's or 2's...
Number of columns with consecutives 1's or 2's >= 10 or 12.
While we wanted to implement these features, we were not sure how to do so in code.